stan/math/fwd/scal/meta/OperandsAndPartials.hpp

```diff
@@ -0,0 +1,188 @@
+     * operands. There are two reason for the generality of this
+     * class. The first is to handle vector and scalar arguments
+     * without needing to write additional code. The second is to use
+     * this class for writing probability distributions that handle
+     * primitives, reverse mode, and forward mode variables
+     * seamlessly.
+     *
+     * This is the partial template specialization for when the return
+     * type is stan::math::fvar<T>.
+     *
+     * @tparam T1 First set of operands.
+     * @tparam T2 Second set of operands.
+     * @tparam T3 Third set of operands.
+     * @tparam T4 Fourth set of operands.
+     * @tparam T5 Fifth set of operands.
+     * @tparam T6 Sixth set of operands.
+     * @tparam T_return_type Return type of the expression. This defaults
+     *   to a template metaprogram that calculates the scalar promotion of
+     *   T1 -- T6.
+     */
+    template<typename T1, typename T2, typename T3,
+             typename T4, typename T5, typename T6,
+             typename T_partials_return>
+    struct OperandsAndPartials<T1, T2, T3, T4, T5, T6,
+                               typename stan::math::fvar<T_partials_return> > {
+      typedef typename stan::math::fvar<T_partials_return> T_return_type;
+
+      const T1& x1_;
+      const T2& x2_;
+      const T3& x3_;
+      const T4& x4_;
+      const T5& x5_;
+      const T6& x6_;
+
+      size_t n_partials;
+      T_partials_return* all_partials;
+
+
+      VectorView<T_partials_return,
+                 is_vector<T1>::value,
+                 is_constant_struct<T1>::value> d_x1;
+      VectorView<T_partials_return,
+                 is_vector<T2>::value,
+                 is_constant_struct<T2>::value> d_x2;
+      VectorView<T_partials_return,
+                 is_vector<T3>::value,
+                 is_constant_struct<T3>::value> d_x3;
+      VectorView<T_partials_return,
+                 is_vector<T4>::value,
+                 is_constant_struct<T4>::value> d_x4;
+      VectorView<T_partials_return,
+                 is_vector<T5>::value,
+                 is_constant_struct<T5>::value> d_x5;
+      VectorView<T_partials_return,
+                 is_vector<T6>::value,
+                 is_constant_struct<T6>::value> d_x6;
+
+      OperandsAndPartials(const T1& x1 = 0, const T2& x2 = 0, const T3& x3 = 0,
+                          const T4& x4 = 0, const T5& x5 = 0, const T6& x6 = 0)
+        : x1_(x1), x2_(x2), x3_(x3), x4_(x4), x5_(x5), x6_(x6),
+          n_partials(!is_constant_struct<T1>::value * length(x1) +
+                     !is_constant_struct<T2>::value * length(x2) +
+                     !is_constant_struct<T3>::value * length(x3) +
+                     !is_constant_struct<T4>::value * length(x4) +
+                     !is_constant_struct<T5>::value * length(x5) +
+                     !is_constant_struct<T6>::value * length(x6)),
+          all_partials(new T_partials_return[n_partials]),
```

Comment:
@bob-carpenter, this used to look like this:
```all_partials(static_cast<T_partials_return*>
                       (vari::operator new
                        (sizeof(T_partials_return) * n_partials)))
```

Should I have left it alone? I was trying to trace back to a way to allocate on the right managed stack without having to touch anything in reverse mode (`stan::vari::operator new` is in reverse mode).